rust/datafusion/src/physical_plan/hash_aggregate.rs

```diff
@@ -214,13 +217,13 @@ Example: average
 * Once all N record batches arrive, `merge` is performed, which builds a RecordBatch with N rows and 2 columns.
 * Finally, `get_value` returns an array with one entry computed from the state
 */
-struct GroupedHashAggregateStream {
-    mode: AggregateMode,
-    schema: SchemaRef,
-    group_expr: Vec<Arc<dyn PhysicalExpr>>,
-    aggr_expr: Vec<Arc<dyn AggregateExpr>>,
-    input: SendableRecordBatchStream,
-    finished: bool,
+pin_project! {
```

Comment:
I found the whole notion of having to Pin things to call `poll_next` on streams to be unnaturally confusing. While I basically understand what `pin_project!` does, I will admit I do not understand why Streams need to be `Pin`. I am just cargo culting it there

Reply:
The crucial observation is that a stream is not an iterator of futures - it is the asynchronous version of an iterator. 

A stream needs to be pinned for the same reason a `Future` [needs to be pinned](https://rust-lang.github.io/async-book/04_pinning/01_chapter.html#why-pinning): the scheduler can stop the work of a `stream` and point to its state via a reference, and the only way for the reference to not become invalid is to forbid the memory from being physically moved (via the `Pin` marker).

Reply:
@jorgecarleitao  -- that is a beautifully eloquent way of explaining the need for Pin. Thank you for that.